### ISSUE TITLE
About Anti-Aliasing

### ISSUE THREAD

**Issue by yumianhuli2** [Unity-Technologies/com.unity.perception#441]
Hello！
    When semantic segmentation(0.9.0 version) is turned on under URP, there is a significant jagged appearance. In Unity's default rendering pipeline, however, jagged is not obvious. Is there a solution?
Thank you!

1、urp unity perception package semantic segmentation alg compare with built-in-RP self semantic segmentation alg 

URP unity perception package semantic segmentation alg（SMAA-HIGH）
![image](https://user-images.githubusercontent.com/39210837/159449831-7f5c0af8-bfa9-4879-9e09-67c45eb98937.png)

built-in-RP self semantic segmentation alg （8X Anti-Aliasing）
![image](https://user-images.githubusercontent.com/39210837/159449850-10d93945-d436-4233-a064-3a0a41ea07c3.png)

2、urp unity perception package rgb compare with urp unity perception package semantic segmentation

unity URP perception package 0.9version RGB 
![image](https://user-images.githubusercontent.com/39210837/159503971-87a2c05b-7722-4248-a0a5-3de13c415357.png)

unity URP perception package 0.9version semantic segmentation
![image](https://user-images.githubusercontent.com/39210837/159503886-71c86432-2f53-41ea-82dc-26946987c861.png)

**Comment by sleal-unity** [comment]
Hi @yumianhuli2,

The jagged appearance (or aliasing) that you see in the semantic segmentation image is actually an intended property of the sort of data segmentation images are meant to capture.

Every pixel of a semantic segmentation image is assigned a uniquely encoded color in order to map which particular class of object occupies each portion of a captured image. For example, the color red might be specifically mapped to the "Car" label, ensuring that all pixels used to rasterize cars in an image to be red. The key thing to note here is that only one unique color identifier can be assigned to an particular pixel in an image. With your typical segmentation image, a mechanism for labeling a pixel as being 50% car and 50% road unfortunately does not exist. If we were to mix two class colors, say red for a car and blue for the road, the resulting color would be purple, but purple in our particular mapping doesn't identify any particular label.

This is why segmentation images can to take on a jagged appearance than their associated RGB images. On the other hand, RGB images get to benefit from the ability to actually blend colors with anti-aliasing algorithms (e.g. FXAA/TAA) and mutli-sampling techniques (e.g. SMAA).

**Comment by yumianhuli2** [comment]
> Hi @yumianhuli2,
> 
> The jagged appearance (or aliasing) that you see in the semantic segmentation image is actually an intended property of the sort of data segmentation images are meant to capture.
> 
> Every pixel of a semantic segmentation image is assigned a uniquely encoded color in order to map which particular class of object occupies each pixel of a captured image. For example, the color red might be specifically mapped to the "Car" label, ensuring that all pixels used to rasterize cars in an image are red. The important detail to note here is that each color is acting as a unique key or identifier for each label class authored in a labeling configuration. With your typical segmentation image, a mechanism for labeling a pixel as being 50% car and 50% road unfortunately does not exist. If we were to mix two class colors, say red for a car and blue for the road, the resulting color would be purple, but purple in our labeling configuration could map to a completely different object class (e.g. sidewalk).
> 
> On the other hand, RGB rasterization gets to benefit from transparent material blending and anti-aliasing techniques (e.g. FXAA, TAA, or SMAA) to produce smoother images. This is why segmentation images can to take on a jagged appearance when compared against their associated RGB images.

ok,got it.Thank you!

**Comment by yumianhuli2** [comment]
> Hi @yumianhuli2,
> 
> The jagged appearance (or aliasing) that you see in the semantic segmentation image is actually an intended property of the sort of data segmentation images are meant to capture.
> 
> Every pixel of a semantic segmentation image is assigned a uniquely encoded color in order to map which particular class of object occupies each pixel of a captured image. For example, the color red might be specifically mapped to the "Car" label, ensuring that all pixels used to rasterize cars in an image are red. The important detail to note here is that each color is acting as a unique key or identifier for each label class authored in a labeling configuration. With your typical segmentation image, a mechanism for labeling a pixel as being 50% car and 50% road unfortunately does not exist. If we were to mix two class colors, say red for a car and blue for the road, the resulting color would be purple, but purple in our labeling configuration could map to a completely different object class (e.g. sidewalk).
> 
> On the other hand, RGB rasterization gets to benefit from transparent material blending and anti-aliasing techniques (e.g. FXAA, TAA, or SMAA) to produce smoother images. This is why segmentation images often have a more jagged appearance than their associated RGB images.

Hey,

You clearly explained how the problem works, but in fact these aliasing brings

very big problem, downstream neural network will also learn the sawtooth shape and affect the training of the next network.

So is there any way to solve this problem?

Thank you！

**Comment by alextha-unity** [comment]
Hello @yumianhuli2 - independent of synthetic/simulated data, a segmentation mask is by definition discrete per-pixel.  In fact, the important part of a segmentation mask is the individual per-pixel class labels which ultimately get converted to tensors, the rendering is just a visualization tool.  This is the intended design for this kind of label in computer vision systems.

The impact of this to any downstream software depends on what you are trying to do with the prediction.  Can you elaborate on a use case of what kind of neural network, or application, you are proposing that would consume the mask?

**Comment by yumianhuli2** [comment]
Hey！

For example, if I input an RGB, it needs to use DDRNET for semantic segmentation. If the truth value is the semantic segmentation graph output by unity Perception package, the network will learn Aliasing, which is quite different from manual labeling.

ddrnet:https://github.com/chenjun2hao/DDRNet.pytorch

Thank you



At 2022-04-07 05:29:25, "Alex Thaman" ***@***.***> wrote:

Hello @yumianhuli2 - independent of synthetic/simulated data, a segmentation mask is by definition discrete per-pixel. In fact, the important part of a segmentation mask is the individual per-pixel class labels which ultimately get converted to tensors, the rendering is just a visualization tool. This is the intended design for this kind of label in computer vision systems.

The impact of this to any downstream software depends on what you are trying to do with the prediction. Can you elaborate on a use case of what kind of neural network, or application, you are proposing that would consume the mask?

**Comment by alextha-unity** [comment]
If you look at the semantic labels in the paper for DDRNET, as an example,
the masks are also quite aliased.  This is a bit harder to observe when the
masks are zoomed out but zooming in you will see the clear jagged edges on
diagonal lines.

There are other label formats for masks such as polygon segmentation which
describe the mask as a polygon.  These do not experience aliasing directly
but projecting onto a 2D image will result in some choice to be made at
each pixel on the boundary.

Are you seeing any particular examples where the mask seems more aliased
than a labeled real world dataset?

**Comment by yumianhuli2** [comment]
oh^_^
